### PR TITLE
Fix problems with UseLayoutRounding.

### LINKED ISSUE #8092: Space appearing in context menus with 150% scaling.
# Describe the bug

Since #7867 space has started appearing in context menus on screens with 150% scaling:

image

There seem to be two issues here:

## ScrollViewer has wrong Viewport

The `ScrollContentPresenter` in the menu's `ScrollViewer` has:

- An `Extent` of `408.6666666666667, 185.33333333333334`
- A `Viewport` of `408.6666666666667, 184.66666666666666`

So the viewport (184) is smaller than the extent (185), meaning that scrollbars get shown. `ScrollViewer`s in menus have a special style, (which are also broken, see the next section), but the "space" being seen here is a `RepeatButton` for scrolling the menu.

This part of the problem was caused by #7867: seems that the `ScrollContentPresenter` size is getting rounded down while its contents are either being rounded up, or not rounded. There is obviously space to display that extra pixel here so this shouldn't be happening.

## Menu scroller style is broken

The extra space being shown is actually a `RepeatButton` from the menu `ScrollViewer` style, but #7735 broke the menu scroller style selector by inserting an additional `Decorator` in the logical tree. See the comments for #7735.

# To Reproduce

1. Open ControlCatalog.NetCore on a screen with 150% scaling
2. Go to ContextMenu tab
3. Open a context menu
4. See space

# Desktop (please complete the following information):**

- OS: All
- Version: master

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

As described in #8092; while #7867 improved our handling of `UseLayoutRounding`, it also introduced a few other problems which were visible with menus on screens with scaling != 1.

This PR tweaks the fixes in #7867 to fix these problems:

- Always round sizes _up_, not to the nearest pixel as before, thereby ensuring that `DesiredSize`s don't get rounded down where possible
- Previously rounding of `Padding` and `BorderThickness` was done in the arrange pass but not the measure pass. Add this rounding to the measure pass as well - this ensures that `DesiredSize` takes this rounding into account
- Fix double comparison in `MenuScrollingVisibilityConverter` to not display scrollbars due to floating point errors

Also adds a few unit tests to test these scenarios.

## Fixed issues

Fixes #8092

### CODE REVIEW

File: src/Avalonia.Base/Point.cs [hunk 1]
```diff
@@ -204,6 +204,18 @@ public bool Equals(Point other)
             // ReSharper enable CompareOfFloatsByEqualityOperator
         }
 
+        /// <summary>
+        /// Returns a boolean indicating whether the point is equal to the other given point
+        /// (numerically).
+        /// </summary>
+        /// <param name="other">The other point to test equality against.</param>
+        /// <returns>True if this point is equal to other; False otherwise.</returns>
+        public bool NearlyEquals(Point other)
```
Comment (grokys) [inline src/Avalonia.Base/Point.cs:213]: This method was already available for `Size` but not for `Point` - it's useful in unit tests if not elsewhere.

**Comment by grokys** [comment]
@AndrejBunjac would be good if you can take a look at this to check that it makes sense!

**Comment by AndrejBunjac** [comment]
Only briefly looked it and the change seems good - I'll take a detailed look over the code tonight after work. :)

Thanks!

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0020407-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Review by AndrejBunjac — APPROVED** [review]
This is long overdue on my end - sorry, I got overwhelmed by other work.
The change looks solid - it didn't even occur to me that DesiredSize could be rounded down but I see how that ends up being an issue.
Good that some unit tests were also added.

**Review by maxkatz6 — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0021405-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by robloo** [comment]
This fixed a few different issues I've seen in control catalog 👍

**Pull request merged**